### PR TITLE
Add Viridis and GrayScale palettes.

### PR DESCRIPTION
**Pull request by ariste-sarl**
Adds the Viridis and GrayScale continuous palettes to the available palettes.
These continuous palettes always distribute the requested colors evenly across the entire color domain.

### CODE REVIEW

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 1]
```diff
@@ -1580,6 +1589,48 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+Description DrawableImplementation::BuildViridisPalette(const UnsignedInteger size)
+{
+  if (IsFirstInitialization)
+  {
+    InitializeValidParameterList();
+    IsFirstInitialization = false;
+  }
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  palette[size - 1] = ViridisColors[255];
```
Comment (mbaudin47) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1603]: Please use `sizeof`.

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 2]
```diff
@@ -1580,6 +1589,48 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+  if (size > 1)
+  {
+    Scalar step = 255. / (size - 1);
+    for (UnsignedInteger i = 0; i + 1 < size; i++)
+      palette[i] = ViridisColors[(UnsignedInteger)(i * step + 0.49)];
```
Comment (mbaudin47) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1608]: Please use `std::round`.

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 3]
```diff
@@ -1580,6 +1589,48 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+{
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  UnsignedInteger firstValue = 255 - std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-WhiteMargin")),
+    lastValue = std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-BlackMargin"));
+  if (size > 1)
+  {
+    Scalar step = (lastValue - (Scalar)firstValue) / (size - 1);
+    for (UnsignedInteger i = 0; i < size; i++){
+      UnsignedInteger val = std::min(255UL, (UnsignedInteger)(firstValue + i * step + 0.49));
```
Comment (mbaudin47) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1625]: Please use `std::round`.

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 4]
```diff
@@ -1580,6 +1589,48 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+      palette[i] = ConvertFromRGB(val,val,val);
```
Comment (mbaudin47) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1626]:
```suggestion
      palette[i] = ConvertFromRGB(val, val, val);
```

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 5]
```diff
@@ -1580,6 +1589,48 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+Description DrawableImplementation::BuildGrayScalePalette(const UnsignedInteger size)
+{
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  UnsignedInteger firstValue = 255 - std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-WhiteMargin")),
+    lastValue = std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-BlackMargin"));
+  if (size > 1)
+  {
+    Scalar step = (lastValue - (Scalar)firstValue) / (size - 1);
+    for (UnsignedInteger i = 0; i < size; i++){
```
Comment (mbaudin47) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1624]:
```suggestion
    for (UnsignedInteger i = 0; i < size; i++)
    {
```

**Comment by josephmure** [comment]
You need to rebase (again) on master to get build_linux and build_mingw to work.

File: python/doc/examples/graphs/plot_graphs_loglikelihood_contour.py [hunk 6]
```diff
@@ -171,21 +171,22 @@ def logLikelihood(X):
 # levels, while others are very different.
 # Anyway, it is obviously nicer to get a colored graphics.
```
Comment (josephmure) [inline python/doc/examples/graphs/plot_graphs_loglikelihood_contour.py:172]: The section title "Multicolor contour plot" and the introduction are not suited to the actual graphs.
I would replace the section title with "Grayscale contour plot" and remove the introductory paragraph.

**Review by josephmure — APPROVED** [review]

**Comment by josephmure** [comment]
@mbaudin47 Could you please take over this PR? A rebase is needed and you still need to do your review, so maybe the best way to proceed is that you create a branch of your own with these developments, do the changes you want, and resubmit it as a new PR.

**Comment by mbaudin47** [comment]
Here the state of the PR.
- A continuous colormap of a contour plot is managed in #2587.
- A discrete colormap is already available in the library.

An example of coloring a set of Curve with a list of Matplotlib colors should be created (e.g. https://openturns.github.io/openturns/latest/auto_functional_modeling/field_functions/plot_viscous_fall_field_function.html).

**Comment by mbaudin47** [comment]
Thank you very much for the good work. I took the PR on my side at #2595. Please close this PR so that the work can continue at #2595.

**Pull request closed without merging**